backend/dataall/modules/dataset_sharing/services/share_managers/lf_share_manager.py

```diff
@@ -432,3 +437,129 @@ def handle_revoke_failure(
+
+    def _verify_catalog_ownership(self, catalog_account_id, catalog_region, catalog_database):
+        """
+        Verifies the catalog ownership by checking
+        1. if the pivot role is assumable in the catalog account
+        2. if "owner_account_id" tag is present in the catalog account, which contains AWS account of source account / producer account -  where the data is present in S3 bucket
+        Returns : Raises exception only in case there is an issue with any of above
+        """
+        if catalog_account_id != self.source_environment.AwsAccountId:
+            logger.info(f'Database {self.dataset.GlueDatabaseName} is a resource link and '
+                        f'the source database {catalog_database} belongs to a catalog account {catalog_account_id}')
+            if SessionHelper.is_assumable_pivot_role(catalog_account_id):
+                self._validate_catalog_ownership_tag(catalog_account_id, catalog_region, catalog_database)
+            else:
+                raise Exception(f'Pivot role is not assumable, catalog account {catalog_account_id} is not onboarded')
+
+    def _validate_catalog_ownership_tag(self, catalog_account_id, catalog_region, catalog_database):
+        glue_client = GlueClient(account_id=catalog_account_id,
+                                 database=catalog_database,
+                                 region=catalog_region)
+
+        tags = glue_client.get_database_tags()
+        if tags.get('owner_account_id', '') == self.source_environment.AwsAccountId:
+            logger.info(
+                f'owner_account_id tag exists and matches the source account id {self.source_environment.AwsAccountId}')
+        else:
+            raise Exception(
+                f'owner_account_id tag does not exist or does not matches the source account id {self.source_environment.AwsAccountId}')
+
+    def check_catalog_account_exists_and_verify(self):
+        """
+        Checks if the source account has a catalog associated with it. This is checked by getting source catalog information and checking if there exists a target database for the source db
+        Return -
+        True - if a catalog account is present and it is verified
+        False - if no source catalog account is present
+        None - if catalog account exists but there is an issue with verifing the conditions needed for source account. Check _verify_catalog_ownership for more details
+        """
+        try:
+            catalog_dict = GlueClient(
+                account_id=self.source_environment.AwsAccountId,
+                region=self.source_environment.region,
+                database=self.dataset.GlueDatabaseName,
+            ).get_source_catalog()
+            if catalog_dict is not None:
+                # Verify the ownership of dataset by checking if pivot role is assumable and ownership tag is present
+                self._verify_catalog_ownership(catalog_dict.get('account_id'), catalog_dict.get('region'),
```

Comment:
Very minor: instead of checking the `if catalog_account_id != self.source_environment.AwsAccountId:` inside `_verify_catalog_ownership`, we could add the condition in line 508: `if catalog_dict is not None and catalog_account_id != self.source_environment.AwsAccountId:`

Reply:
What benefit would that provide ? 

So are you suggesting from the point of view that when a catalog_dict is formulated the responsiblity of `_verify_catalog_ownership` remains only to check if pivot role is assumable and that the tags are present

Reply:
It is more from a readability POV. My first thought when I saw the function was "wait are we verifying the ownership for the case environment account = catalog account?" But it is very minor, one can go to the function and see that we only verify when the accounts are different